File: src/utils/fleetCalculations.ts
import { Vessel } from '../types/maritime';

export function healthBand(score: number): 'Healthy' | 'Warning' | 'Critical' {
  if (score >= 80) return 'Healthy';
  if (score >= 60) return 'Warning';
  return 'Critical';
}

export function fleetKpis(vessels: Vessel[]): Record<string, string | number> {
  const delayed = vessels.filter((v) => {
    const isDelayedEta = new Date(v.predicted_eta) > new Date(v.planned_eta);
    return isDelayedEta || v.voyage_status === 'Delayed';
  }).length;

  const underway = vessels.filter((v) => v.operational_status === 'Underway').length;
  const inPort = vessels.filter((v) => v.operational_status === 'In Port').length;
  const criticalRisk = vessels.filter((v) => v.safety_risk_level === 'Critical').length;
  const avgHealth =
    vessels.length > 0
      ? vessels.reduce((acc, v) => acc + (v.technical_health_score || 0), 0) / vessels.length
      : 0;

  return {
    'Total vessels': vessels.length,
    'Vessels underway': underway,
    'Vessels in port': inPort,
    'Delayed vessels': delayed,
    'Critical risk': criticalRisk,
    'Average fleet health': avgHealth.toFixed(1),
  };
}

File: legacy/streamlit/utils/fleet_calculations.py
"""Transparent summary rules for the synthetic fleet module."""

from __future__ import annotations

import pandas as pd


def fleet_kpis(vessels: pd.DataFrame) -> dict[str, str | int]:
    delayed = (vessels["predicted_eta"] > vessels["planned_eta"]) | (vessels["voyage_status"] == "Delayed")
    return {
        "Total vessels": len(vessels),
        "Vessels underway": int((vessels["operational_status"] == "Underway").sum()),
        "Vessels in port": int((vessels["operational_status"] == "In Port").sum()),
        "Delayed vessels": int(delayed.sum()),
        "Critical risk": int((vessels["safety_risk_level"] == "Critical").sum()),
        "Average fleet health": f"{vessels['technical_health_score'].mean():.1f}",
    }


def health_band(score: float) -> str:
    if score >= 80:
        return "Healthy"
    if score >= 60:
        return "Warning"
    return "Critical"

